### ISSUE TITLE
How to deal with select in findOne so that inferred entity is partial?

### ISSUE THREAD

**Issue by nicgirault** [typeorm/typeorm#4113]
**Issue type:**

[x] question
[ ] bug report
[ ] feature request
[ ] documentation issue

**TypeORM version:**

[x] `latest`
[ ] `@next`
[ ] `0.x.x` (or put your version here)

One of the feature I'm looking for is the following:

```typescript
const user = await User.findOne(userId, { select: ['firstName'] })

if (!user) {
  return
}

console.log(user.lastName)
```

I would expect typescript to warn because I didn't select `lastName` but it doesn't. Is there a way to be typesafe with `select`?

**Comment by Kononnable** [comment]
I think the only way is to cast it manually. 
It would be very hard(if this is even possible) to make find functions with select options truly type-safe. Always returning partial wouldn't be a good solution since most of the times select option isn't passed and it would require everyone to check for each property is it undefined before using it.

**Comment by ifeltsweet** [comment]
Prisma [does this](https://www.prisma.io/docs/reference/tools-and-interfaces/prisma-client/field-selection) and it does it [exceptionally well](https://www.youtube.com/watch?v=LggrE5kJ75I). I believe it's possible in Prisma because the client is generated from your schema.

**Comment by TomCaserta** [comment]
I don't think you need to generate anything here, for example I believe if 'findOne/findOneOrFail' had the following signature (or similar, this is just the 'quickest' solution I could come up with):

```
findOneOrFail<T extends Array<keyof Entity>>(id?: string | number | Date | ObjectID, options?: FindOneOptions<Entity> & { select: T }): Promise<Pick<Entity, T[number]>>;
```

Then this would mean that only keys you specify inside of `select` are available on the returned entity.

**Comment by ifeltsweet** [comment]
I believe currently this is possible only for the first level of keys. If you need to drill down into a nested object you won't be able to do it with the way selection currently works because it uses a string with dot notation, eg: `"address.zip"`.

If selects were done with an object like:

```
{
  address: {
    zip: 1
  }
}
```

then it would be possible to do it without code generation. I have actually implemented a similar wrapper for Mongoose queries that converts selection object as shown above into string dot notation that MongoDB understands:

```
export function select<T, S extends StrictSelection<Unpacked<T>, S>>(
  query: Query<T>,
  selection: S,
): Query<T extends any[] ? Select<Unpacked<T>, S>[] : Select<Unpacked<T>, S>> {
  return query.select(getSelection(selection)) as any;
}

type Select<T, S extends Selection<T>> = DeepPick<T, S> & Omit<T, keyof DocumentDefinition<T>> & { id: any };
type Selection<T> = { [K in keyof T]?: 0 | 1 | Selection<T[K]> };
type StrictSelection<T, S> = { [K in keyof S]?: K extends keyof T ? 0 | 1 | StrictSelection<T[K], S[K]> : never };

type DeepPick<T, S extends Selection<T>> = OmitProperties<{
  [K in keyof S]: K extends keyof T
  ? S[K] extends 1
  ? T[K]
  : S[K] extends 0
  ? never
  : S[K] extends Selection<T[K]>
  ? DeepPick<T[K], S[K]>
  : never
  : never
}, never>;

type Unpacked<T> = T extends (infer U)[] ? U : T;
type OmitProperties<T, P> = Pick<T, { [K in keyof T]: T[K] extends P ? never : K }[keyof T]>;
```

**Comment by derekparsons718** [comment]
This seems like a big hole in the type system. Explicit casting is quite messy and far from ideal. It would be awesome to see this improved.
